packages/wavs/src/subsystems/trigger/streams/evm_stream/client/connection.rs

```diff
@@ -524,4 +749,107 @@ mod test {
+            index: 1,
+            chain_key: wavs_types::ChainKey::new("evm:31337").expect("Invalid chain key format"),
+        };
+
+        let connection = Connection::new(
+            rpc_ids,
+            endpoints,
+            channels.connection,
+            Some(priority_endpoint),
+        );
+
+        // Step 1: Should connect to priority endpoint (anvil_2) first
+        let result = timeout(Duration::from_secs(10), async {
+            loop {
+                if let Some(endpoint) = connection.current_endpoint() {
+                    if endpoint == anvil_2.ws_endpoint() {
+                        tokio::time::sleep(Duration::from_millis(100)).await;
+                        if connection.current_endpoint().as_ref() == Some(&endpoint) {
+                            return endpoint;
+                        }
+                    }
+                }
+                tokio::time::sleep(Duration::from_millis(50)).await;
+            }
+        })
+        .await;
+
+        let initial_connection = result.expect("Should connect to priority endpoint first");
+        assert_eq!(anvil_2.ws_endpoint(), initial_connection);
+        tracing::info!("✓ Connected to priority endpoint (anvil_2) first");
+
+        // Step 2: Drop priority endpoint to simulate connection loss
+        let anvil_2_port = anvil_2.port();
+        drop(anvil_2);
+        tokio::time::sleep(Duration::from_millis(200)).await;
+
+        // Should fallback to anvil_1
+        let result = timeout(Duration::from_secs(10), async {
+            loop {
+                if let Some(endpoint) = connection.current_endpoint() {
+                    if endpoint == anvil_1.ws_endpoint() {
+                        tokio::time::sleep(Duration::from_millis(100)).await;
+                        if connection.current_endpoint().as_ref() == Some(&endpoint) {
+                            return endpoint;
+                        }
+                    }
+                }
+                tokio::time::sleep(Duration::from_millis(50)).await;
+            }
+        })
+        .await;
+
+        let fallback_connection = result.expect("Should fallback to non-priority endpoint");
+        assert_eq!(anvil_1.ws_endpoint(), fallback_connection);
+        tracing::info!(
+            "✓ Fell back to non-priority endpoint: {}",
+            fallback_connection
+        );
+
+        // Step 3: Recreate priority endpoint and verify reconnection
+        let anvil_2_recreated = Anvil::new().port(anvil_2_port).spawn();
+        tokio::time::sleep(Duration::from_millis(100)).await;
+
+        // Wait for health check to detect the priority endpoint and force reconnection
+        let result = timeout(Duration::from_secs(15), async {
+            loop {
+                if let Some(endpoint) = connection.current_endpoint() {
+                    if endpoint == anvil_2_recreated.ws_endpoint() {
+                        tokio::time::sleep(Duration::from_millis(200)).await;
```

Comment:
why do we need this additional inner loop?

Reply:
oh, so it's checking that it didn't change _again_ after the 200ms wait. Could help to have a comment there since 200ms is a pretty small window for things to change, but no biggie